DEVX-156 add docs for state sync (#1044)
commercetools/commercetools-sync-java: docs/sdk2/usage/STATE_SYNC.md

    final Function<LocalizedString, StateUpdateAction> createBeforeUpdateAction =
        (newName) -> StateSetNameAction.builder().name(newName).build();

// Add the change name action to the list of update actions before update is executed
    final TriFunction<
            List<StateUpdateAction>, StateDraft, State, List<StateUpdateAction>>
        beforeUpdateStateCallback =
            (updateActions, newStateDraft, oldState) -> {
              final StateUpdateAction beforeUpdateAction =
                  createBeforeUpdateAction.apply(newStateDraft.getName());
              updateActions.add(beforeUpdateAction);
              return updateActions;
            };
```

### Build StateDraft (syncing from external project)

The state-sync expects a list of `StateDraft`s to process. If you use java-sync-library to sync your states from any external system into a commercetools platform project you have to convert your data into CTP compatible `StateDraft` type. This was done in previous version using `DraftBuilder`s.
The V2 SDK do not have inheritance for `DraftBuilder` classes but the differences are minor and you can replace it easily. Here's an example:

```java

> // StateDraftBuilder in v1 takes parameters 'name', 'cartPredicate', 'value', 'target', 'sortOrder', 'requiresDiscountCode' 

Reviewer: Adjust the comment

Author: https://github.com/commercetools/commercetools-sync-java/pull/1044/commits/a3b6146ca828af2079f8991e138b8ce5124dce3e

|Draft|Required Fields|Note|
|---|---|---|
| [StateDraft](https://docs.commercetools.com/api/projects/states#statedraft)| `key` |  Also, the states in the target project are expected to have the `key` fields set. | 

#### Reference Resolution 

`Transitions` are a way to describe possible transformations of the current state to other states of the same type (for example Initial -> Shipped). When performing a [SetTransitions](https://docs.commercetools.com/api/projects/states#set-transitions), an array of [ResourceIdentifiers](https://docs.commercetools.com/api/projects/states#ctp:api:type:StateResourceIdentifier) to State is needed.
In commercetools, a [ResourceIdentifier](https://docs.commercetools.com/api/projects/states#ctp:api:type:StateResourceIdentifier) can be created by providing either the key or the ID. 
When the key to a referenced state is provided with a `ResourceIdentifier`, the sync will resolve the resource with the given key and use the ID of the found resource to create or update a reference. 

| ResourceIdentifier Field | Type                                       |
|:-------------------------|:-------------------------------------------|

> | `transitions`            | Array of StateResourceIdentifiers to State | 

Reviewer: ```suggestion
| `transitions`            | Array of StateResourceIdentifiers | 
```

* sync exception
* state draft from the source
* state of the target project (only provided if an existing state could be found)
* the update-actions, which failed (only provided if an existing state could be found)

````java
 final Logger logger = LoggerFactory.getLogger(StateSync.class);
 final StateSyncOptions stateSyncOptions = StateSyncOptionsBuilder
         .of(projectApiRoot)
         .errorCallback((syncException, draft, state, updateActions) -> 
            logger.error(new SyncException("My customized message"), syncException)).build();
````
    
##### warningCallback

A callback is called whenever a warning event occurs during the sync process. Each resource executes its own 
warning-callback. When the sync process of a particular resource runs successfully, it is not triggered. It contains the 
following context about the warning message:

* sync exception
* state draft from the source 
* state of the target project (only provided if an existing state could be found)

````java
 final Logger logger = LoggerFactory.getLogger(StateSync.class);
 final StateSyncOptions stateSyncOptions = StateSyncOptionsBuilder
         .of(projectApiRoot)

>          .warningCallback((syncException, draft, state, updateActions) -> 

Reviewer: ```suggestion
         .warningCallback((syncException, draft, state) -> 
```

            logger.warn(new SyncException("My customized message"), syncException)).build();
````

##### beforeUpdateCallback
During the sync process, if a target state and a state draft are matched, this callback can be used to 
intercept the **_update_** request just before it is sent to the commercetools platform. This allows the user to modify 
update actions array with custom actions or discard unwanted actions. The callback provides the following information :
 
 * state draft from the source
 * state from the target project
 * update actions that were calculated after comparing both

````java
final TriFunction<
        List<StateUpdateAction>, StateDraft, State, List<StateUpdateAction>> 
            beforeUpdateStateCallback =
            (updateActions, newStateDraft, oldState) ->  updateActions.stream()
                    .filter(updateAction -> !(updateAction instanceof StateRemoveRolesAction))
                    .collect(Collectors.toList());
                        
final StateSyncOptions stateSyncOptions = 

>         StateSyncOptionsBuilder.of(sphereClient).beforeUpdateCallback(beforeUpdateStateCallback).build();

Reviewer: ```suggestion
        StateSyncOptionsBuilder.of(projectApiRoot).beforeUpdateCallback(beforeUpdateStateCallback).build();
```

````

##### beforeCreateCallback
During the sync process, if a state draft should be created, this callback can be used to intercept the **_create_** request just before it is sent to the commercetools platform.  It contains the following information : 

 * state draft that should be created

Please refer to [example in product sync document](PRODUCT_SYNC.md#example-set-publish-stage-if-category-references-of-given-product-draft-exists).

##### batchSize
A number that could be used to set the batch size with which states are fetched and processed,
as states are obtained from the target project on commercetools platform in batches for better performance. The algorithm accumulates up to `batchSize` resources from the input list, then fetches the corresponding states
from the target project on the commecetools platform in a single request. Playing with this option can slightly improve or reduce processing speed. If it is not set, the default batch size is 50 for state sync.

````java                         
final StateSyncOptions stateSyncOptions = 

>          StateSyncOptionsBuilder.of(sphereClient).batchSize(30).build();

Reviewer: ```suggestion
         StateSyncOptionsBuilder.of(projectApiRoot).batchSize(30).build();
```

It being persisted as `CustomObject` means that the referenced StateDrafts with keys `state-B`  do not exist yet.

```json
{
  "container": "commercetools-sync-java.UnresolvedTransitionsService.stateDrafts",
  "key": "518ea82bb78755c0cdd67909dd3206d56186f7e5",
  "value": {
    "missingTransitionStateKeys": [
      "state-B"
    ],
    "stateDraft": {
      "type": "ReviewState",
      "transitions": [
        {
          "id": "state-B",
          "typeId": "state"
        }
      ],
      "roles": [
        "ReviewIncludedInStatistics"
      ],
      "key": "state-A",
      "initial": true
    }
  }
}
```

As soon, as the referenced StateDrafts are supplied to the sync, the draft will be created/updated and the 
`CustomObject` will be removed from the target project.

Keeping the old custom objects around forever can negatively influence the performance of your project and the time it takes to restore it from a backup. Deleting unused data ensures the best performance for your project. Please have a look into the [Cleanup guide](CLEANUP_GUIDE.md) to cleanup old unresolved custom objects.

#### More examples of how to use the sync
 

>  1. [Sync usages](https://github.com/commercetools/commercetools-sync-java/tree/master/src/integration-test/java/com/commercetools/sync/integration/externalsource/states/StateSyncIT.java).

Reviewer: ```suggestion
 1. [Sync usages](#todo).
```

*Make sure to read the [Important Usage Tips](IMPORTANT_USAGE_TIPS.md) for optimal performance.*

### Build all update actions

A utility method provided by the library to compare a `State` with a new `StateDraft` and results in a list of state update actions.
 update actions. 
```java
List<StateUpdateAction> updateActions = StateSyncUtils.buildActions(state, stateDraft, stateSyncOptions);
```

### Build particular update action(s)

Utility methods provided by the library to compare the specific fields of a `State` and a new `StateDraft`, and in turn builds
 the update action. One example is the `buildSetNameAction` which compares names:
````java
Optional<StateUpdateAction> updateAction = StateUpdateActionUtils.buildSetNameAction(oldState, stateDraft);
````

> More examples of those utils for different types can be found [here](https://github.com/commercetools/commercetools-sync-java/tree/master/src/test/java/com/commercetools/sync/states/utils/StateUpdateActionUtilsTest.java).

Reviewer: ```suggestion
More examples of those utils for different types can be found [here](#todo).
```

## Migration Guide

> The cart-discount-sync uses the [JVM-SDK-V2](http://commercetools.github.io/commercetools-sdk-java-v2), therefore ensure you [Install JVM SDK](https://docs.commercetools.com/sdk/java-sdk-getting-started#install-the-java-sdk) module `commercetools-sdk-java-api` with

Reviewer: ```suggestion
The state-sync uses the [JVM-SDK-V2](http://commercetools.github.io/commercetools-sdk-java-v2), therefore ensure you [Install JVM SDK](https://docs.commercetools.com/sdk/java-sdk-getting-started#install-the-java-sdk) module `commercetools-sdk-java-api` with
```
